src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,514 @@
+            return newSize;
+        }
+
+        public TValue GetAt(int index) => GetAt(index, out _);
+
+        public TValue GetAt(int index, out TKey key)
+        {
+            if (index < 0 || index >= Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), "Index was out of range. Must be non-negative and less than the size of the collection.");
+            }
+
+            Entry entry = _entries[index];
+            key = entry.Key;
+            return entry.Value;
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(this);
+
+        public TValue GetValue(TKey key)
+        {
+            int index = IndexOf(key);
+            if (index < 0)
+            {
+                throw new KeyNotFoundException("The given key was not present in the dictionary.");
+            }
+            return _entries[index].Value;
+        }
+
+        public int IndexOf(TKey key) => IndexOf(key, out _);
+
+        public void Insert(int index, TKey key, TValue value)
+        {
+            if (index < 0 || index > Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), "Index was out of range. Must be non-negative and less than or equal to the size of the collection.");
+            }
+
+            TryInsert(index, key, value, InsertionBehavior.ThrowOnExisting);
+        }
+
+        public bool Remove(TKey key) => Remove(key, out _);
+
+        public bool Remove(TKey key, out TValue value)
+        {
+            int index = IndexOf(key);
+            if (index >= 0)
+            {
+                value = _entries[index].Value;
+                RemoveAt(index);
+                return true;
+            }
+            value = default;
+            return false;
+        }
+
+        public void RemoveAt(int index)
+        {
+            int count = Count;
+            if (index < 0 || index >= count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), "Index was out of range. Must be non-negative and less than the size of the collection.");
+            }
+
+            // Remove the entry from the bucket
+            UpdateBucket(index, true);
+
+            // Decrement the indices > index
+            for (int i = index + 1; i < count; ++i)
+            {
+                UpdateBucket(i, false, -1);
+            }
+            Entry[] entries = _entries;
+            Array.Copy(entries, index + 1, entries, index, count - index - 1);
+            --Count;
+            entries[Count] = default;
+            ++_version;
+        }
+
+        public void SetAt(int index, TValue value)
+        {
+            if (index < 0 || index >= Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), "Index was out of range. Must be non-negative and less than the size of the collection.");
+            }
+
+            _entries[index].Value = value;
+            ++_version;
```

Comment:
We could choose to avoid `++_version` for `SetAt()` and for `SetValue()` which ended up updating an existing value, because that will not disrupt enumeration. I am not sure why Dictionary increments it in that case.